scripts/check.sh

```diff
@@ -0,0 +1,153 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Checks that the current state of the tree is sane and optionally auto-fixes
+# errors automatically. Meant for interactive use.
+
+function usage() {
+  cat <<EOF
+USAGE: scripts/check.sh [--allow-dirty] [--commit] [<revision>]
+
+Runs auto-formatting scripts, source-tree checks, and linters on any files that
+have changed since master.
+
+By default, any changes are left as uncommited changes in the working tree. You
+can review them with git diff. Pass --commit to automatically commit any changes.
+
+Pass an alternate revision to use as the basis for checking changes.
+
+OPTIONS:
+
+  --allow-dirty
+    By default, check.sh requires a clean working tree to keep any generated
+    changes separate from logical changes.
+
+  --commit
+    Commit any auto-generated changes.
+
+  <revision>
+    Specifies a starting revision other than the default of master.
+
+
+EXAMPLES:
+
+  check.sh
+    Runs automated checks and formatters on all changed files since master.
+    Check for changes with git diff.
+
+  check.sh --commit
+    Runs automated checks and formatters on all changed files since master and
+    commits the results.
+
+  check.sh --commit HEAD
+    Runs automated checks and formatters on all changed files since the last
+    commit.
+
+  check.sh --allow-dirty HEAD
+    Runs automated checks and formatters on all changed files since the last
+    commit and intermingles the changes with any pending changes. Useful for
+    interactive use from an editor.
+
+EOF
+}
+
+set -euo pipefail
+unset CDPATH
+
+# Change to the top-directory of the working tree
+top_dir=$(git rev-parse --show-toplevel)
+cd "$top_dir"
+
+allow_dirty=false
+commit=false
+start="master"
+
+while [[ $# -gt 0 ]]; do
+  case "$1" in
+    --)
+      # Do nothing: explicitly allow this, but ignore it
+      ;;
+
+    -h | --help)
+      usage
+      exit 1
+      ;;
+
+    --allow-dirty)
+      allow_dirty=true
+      ;;
+
+    --commit)
+      commit=true
+      ;;
+
+    *)
+      if git rev-parse "$1" >& /dev/null; then
+        start="$1"
+        break
+      fi
+      ;;
+  esac
+  shift
+done
+
+if [[ $allow_dirty == true && $commit == true ]]; then
+  echo "--allow_dirty and --commit are mutually exclusive"
+  exit 1
+fi
+
+if ! git diff-index --quiet HEAD --; then
+  if [[ $allow_dirty != true ]]; then
+    echo "You have local changes that could be overwritten by this script."
+    echo "Please commit your changes first or pass --allow-dirty."
+    exit 2
+  fi
+fi
+
+# Record actual start
+start_sha=$(git rev-parse "$start")
+
+# Restyle and commit any changes
+scripts/style.sh "$start_sha" || :
+if ! git diff --quiet; then
+  if [[ $commit == true ]]; then
+    echo "Style generated changes"
+    git commit -a --fixup=HEAD
```

Comment:
I guess this is fine. If your PR is all in a single commit then this seems exactly correct. But if your PR contains multiple commits, this will effectively "amend/squash" (though only after doing a rebase) with the last commit, which isn't necessarily where the mis-formats were introduced. That's not quite correct, but it's also not wrong enough to matter.

An alternative would be `git commit -a -m "style.sh"` or similar. That won't auto-squash during a rebase, but that might be viewed as a feature in this case.

No action required. (Though if you change anything here, also do so below).

Reply:
Ah, so you'd run check.sh for *each* commit instead of just once at the end of the chain. Yeah, that'd work perfectly. Having these options is even better though. :)